src/main/java/com/arpnetworking/utility/RegexAndMapReplacer.java

```diff
@@ -0,0 +1,296 @@
+            }
+        }
+    }
+
+    private static String replacePrefix(
+            final Matcher matcher,
+            final LinkedHashMap<String, Map<String, String>> variables,
+            final ImmutableList.Builder<String> variablesUsedBuilder,
+            final String prefix,
+            final String variableName) {
+        if (prefix.equals("capture")) {
+            if (isNumeric(variableName)) {
+                final Integer groupIndex = Integer.valueOf(variableName);
+                return matcher.group(groupIndex);
+            } else {
+                try {
+                    return matcher.group(variableName);
+                } catch (final IllegalArgumentException ignored2) { // No group with this name
+                    return "";
+                }
+            }
+        }
+
+        // Only record variables that are not captures
+        variablesUsedBuilder.add(String.format("%s:%s", prefix, variableName));
+
+        final Map<String, String> variableMap = variables.get(prefix);
+        if (variableMap == null) {
+            throw new IllegalArgumentException(String.format("could not find map for variables with prefix '%s'", prefix));
+        }
+        return variableMap.getOrDefault(variableName, "");
+    }
+
+    private static String replaceNonPrefix(
+            final Matcher matcher,
+            final LinkedHashMap<String, Map<String, String>> variables,
+            final ImmutableList.Builder<String> variablesUsedBuilder,
+            final String variableName) {
+        // First try the capture group with the name
+        try {
+            return matcher.group(variableName);
+        } catch (final IllegalArgumentException e) { // No group with this name
+            // Walk through the variable maps in order to find the first match
+            for (final Map.Entry<String, Map<String, String>> entry : variables.entrySet()) {
+                final Map<String, String> variableMap = entry.getValue();
+                final String replacement = variableMap.get(variableName);
+                if (replacement != null) {
+                    variablesUsedBuilder.add(String.format("%s:%s", entry.getKey(), variableName));
+                    return replacement;
+                }
+            }
+        }
+        return "";
+    }
+
+    private static boolean isNumeric(final String string) {
+        boolean isNumeric = true;
+        for (int x = 0; x < string.length(); x++) {
+            if (!Character.isDigit(string.charAt(x))) {
+                isNumeric = false;
+                break;
+            }
+        }
+        return isNumeric;
+    }
+
+    private RegexAndMapReplacer() { }
+
+    /**
+     * Describes the replacement string and variables used in it's creation.
+     *
+     * The "replacement" field is the resulting string.
+     * The "variablesMatched" field is a list of input variables that were matched, in prefix:variable form
```

Comment:
I'm not nuts about maintaining the internal data structures with the prefixes. The reason is that it means parsing and escaping the colons in more than one place. Have you considered just using the prefix to refer to a different input map? (e.g. map of maps where first key is prefix -- for example)

Reply:
That's fair. I'll return that same map structure as gets passed in.